// strongloop-license/lib/handlers.js
exports.CONSOLE = logIfFailure;
exports.EXIT = exitIfFailure;
exports.NOOP = noop;

// Handlers are given:
//  err: null or Error
//  req: "product:feature" or "product" used in license requirement
//  res: true if any license met the requirements

function noop() {}

function exitIfFailure(err, req, res) {
  logIfFailure(err, req, res);
  if (err || !res) {
    console.error('##  Terminating process.');
    // Delay actual exit in case stderr is non-blocking
    setTimeout(exit, 50);
    function exit() {
      process.exit(1);
    }
  }
}

function logIfFailure(err, req, res) {
  if (err || !res) {
    console.error('## %s licensing missing or invalid. Please verify your ' +
                  'licenses in the Licenses page by running StrongLoop Arc ' +
                  'by typing "slc arc --licenses" . If you have questions ' +
                  'about your license or StrongLoop licensing please ' +
                  'contact sales@strongloop.com.', req);
  } else {
    console.log('## %s is licensed from %s to %s.', req,
                res.activationDate.toISOString(),
                res.expirationDate.toISOString());
  }
}
