// strong-trace/lib/waterfall.js
"use strict";

module.exports = Waterfall

var hrt = require("./hrtutils")

function Waterfall(tZero) {
  this.segments = []
  this.blockCache = {}
  this.blockNames = []
  this.tZero = tZero
  this.tags = []
}

Waterfall.prototype.addSegment = function addSegment(block) {
  var i
  var blockId = block.tZero[0] + ":" + block.tZero[1]
  if (this.blockCache[blockId] !== undefined) {
    // we've 'seen' it.
    return true
  }
  this.blockCache[blockId] = true
  this.blockNames.push(block.name)
  for (i = 0; i < block.nodes.length; i++) {
    var node = block.nodes[i]
    //console.log(node)
    var start = hrt.diffMicros(this.tZero, node.startTs)
    var end = hrt.diffMicros(this.tZero, node.endTs)
    this.segments.push(new Segment("fn", node.id, start, end, node.visits))
  }
  for (i = 0; i < block.tags.length; i++) {
    this.tags.push(block.tags[i])
  }
}

Waterfall.prototype.addLink = function addLink(schedule, resume) {
  var start = hrt.diffMicros(this.tZero, schedule.startTs)
  var end = hrt.diffMicros(this.tZero, resume.endTs)
  var waitSeg = new Segment("wait", schedule.name, start, end, 1)
  waitSeg.source = schedule.fn
  waitSeg.target = resume.fn
  this.segments.push(waitSeg)
}

function Segment(type, id, start, end, visits) {
  this.type = type
  this.id = id
  this.start = start
  this.end = end
  this.visits = visits
}
Segment.prototype.source = undefined
Segment.prototype.target = undefined
